src/frontend/Ast_to_Mir.ml

```diff
@@ -520,14 +529,44 @@ let trans_fun_def declc (ts : Ast.typed_statement) =
   in
   stmt
 
-let trans_prog filename
-    { Ast.functionblock
-    ; datablock
-    ; transformeddatablock
-    ; parametersblock
-    ; transformedparametersblock
-    ; modelblock
-    ; generatedquantitiesblock } : typed_prog =
+let gen_write decl_id sizedtype =
+  let bodyfn var =
+    { stmt=
+        NRFunApp (CompilerInternal, string_of_internal_fn FnWriteParam, [var])
+    ; smeta= no_span }
+  in
+  for_scalar sizedtype bodyfn
+    { expr= Var decl_id
+    ; emeta= {internal_meta with mtype= remove_size sizedtype} }
+    no_span
+
+let gen_writes block_filter vars =
+  List.filter_map
+    ~f:(function
+      | decl_id, (st, block) when block = block_filter ->
+          Some (gen_write decl_id st)
+      | _ -> None)
+    vars
+
+let compiler_if compiler_internal_var stmts =
+  let body =
+    match stmts with
+    | [({stmt= Block _; _} as s)] -> s
+    | ls -> {stmt= Block ls; smeta= no_span}
+  in
+  let cond = {expr= Var compiler_internal_var; emeta= internal_meta} in
+  match stmts with
+  | [] -> []
+  | _ -> [{stmt= IfElse (cond, body, None); smeta= no_span}]
+
+let get_block block prog =
+  match block with
+  | Parameters -> prog.Ast.parametersblock
+  | TransformedParameters -> prog.transformedparametersblock
+  | GeneratedQuantities -> prog.generatedquantitiesblock
+  | Data -> prog.datablock
+
+let trans_prog filename p : typed_prog =
```

Comment:
I really like the structure of `trans_prog` now, with these `Check` and `Constrain` and `Read` actions. I wonder if it's worth @bob-carpenter having a look at just this function as a sanity check for the high-level structure of what needs to happen.